File: engines/python/archscope_engine/models/thread_snapshot.py
"""Language-agnostic thread-dump record models (T-188).

The legacy ``ThreadDumpRecord`` (in :mod:`models.thread_dump`) stays the
canonical type for the Java single-dump path. The multi-language /
multi-dump pipeline added in Phase 5 (Java jstack, Go goroutine, Python
py-spy / faulthandler, Node.js diagnostic report, .NET dump) uses the
records defined here so the correlator does not need to know about any
particular runtime.

Three layers:

* :class:`StackFrame` — one stack frame, normalized across runtimes. The
  ``language`` field lets language-aware enrichment plugins (T-194/T-195
  for Java, etc.) opt in or out without inspecting frame text.
* :class:`ThreadSnapshot` — a single thread captured at a single moment.
* :class:`ThreadDumpBundle` — all snapshots from a single dump file,
  carrying provenance (dump index, captured timestamp, originating file,
  detected source format).
"""
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import Any


class ThreadState(str, Enum):
    """Normalized thread states across runtimes.

    Subclasses ``str`` so the enum values serialize naturally to JSON
    (``"RUNNABLE"``) and can be compared against parser strings without an
    explicit ``.value``.
    """

    RUNNABLE = "RUNNABLE"
    BLOCKED = "BLOCKED"
    WAITING = "WAITING"
    TIMED_WAITING = "TIMED_WAITING"
    NETWORK_WAIT = "NETWORK_WAIT"
    IO_WAIT = "IO_WAIT"
    LOCK_WAIT = "LOCK_WAIT"
    CHANNEL_WAIT = "CHANNEL_WAIT"
    DEAD = "DEAD"
    NEW = "NEW"
    UNKNOWN = "UNKNOWN"

    @classmethod
    def coerce(cls, value: str | None) -> "ThreadState":
        """Best-effort mapping of a raw runtime state string to the enum."""
        if value is None:
            return cls.UNKNOWN
        upper = value.strip().upper().replace("-", "_").replace(" ", "_")
        # Common aliases observed in the wild.
        aliases = {
            "RUNNING": cls.RUNNABLE,
            "RUNNABLE": cls.RUNNABLE,
            "ACTIVE": cls.RUNNABLE,
            "BLOCKED": cls.BLOCKED,
            "WAIT": cls.WAITING,
            "WAITING": cls.WAITING,
            "TIMED_WAITING": cls.TIMED_WAITING,
            "PARKED": cls.WAITING,
            "TERMINATED": cls.DEAD,
            "DEAD": cls.DEAD,
            "NEW": cls.NEW,
            "INITIALIZING": cls.NEW,
            "SLEEP": cls.TIMED_WAITING,
            "SLEEPING": cls.TIMED_WAITING,
            "IO_WAIT": cls.IO_WAIT,
            "IO": cls.IO_WAIT,
            "NETWORK_WAIT": cls.NETWORK_WAIT,
            "LOCK_WAIT": cls.LOCK_WAIT,
            "CHANNEL_WAIT": cls.CHANNEL_WAIT,
            "CHAN_RECEIVE": cls.CHANNEL_WAIT,
            "CHAN_SEND": cls.CHANNEL_WAIT,
            "SELECT": cls.CHANNEL_WAIT,
        }
        if upper in aliases:
            return aliases[upper]
        try:
            return cls(upper)
        except ValueError:
            return cls.UNKNOWN


@dataclass(frozen=True)
class StackFrame:
    """A single normalized stack frame.

    ``function`` is the only required field; the rest are best-effort
    captures of whatever the source format provided so language-specific
    enrichment plugins can decide how much detail is available.
    """

    function: str
    module: str | None = None
    file: str | None = None
    line: int | None = None
    language: str | None = None

    def to_dict(self) -> dict[str, Any]:
        return {
            "function": self.function,
            "module": self.module,
            "file": self.file,
            "line": self.line,
            "language": self.language,
        }

    def render(self) -> str:
        """Human-readable single-line rendering used for stack signatures."""
        parts: list[str] = []
        if self.module:
            parts.append(f"{self.module}.{self.function}")
        else:
            parts.append(self.function)
        if self.file:
            location = self.file
            if self.line is not None:
                location = f"{self.file}:{self.line}"
            parts.append(f"({location})")
        return " ".join(parts)


@dataclass
class ThreadSnapshot:
    """One thread captured in one dump."""

    snapshot_id: str
    thread_name: str
    thread_id: str | None
    state: ThreadState
    category: str | None
    stack_frames: list[StackFrame] = field(default_factory=list)
    lock_info: str | None = None
    metadata: dict[str, Any] = field(default_factory=dict)
    language: str | None = None
    source_format: str | None = None

    def stack_signature(self, depth: int = 5) -> str:
        """Compact representation of the top ``depth`` frames.

        Matches the convention used by the legacy single-dump analyzer
        (`" | "` separator) so existing dashboards keep grouping the same
        threads even when they switch to the multi-dump pipeline.
        """
        if not self.stack_frames:
            return "(no-stack)"
        return " | ".join(frame.render() for frame in self.stack_frames[:depth])

    def to_dict(self) -> dict[str, Any]:
        return {
            "snapshot_id": self.snapshot_id,
            "thread_name": self.thread_name,
            "thread_id": self.thread_id,
            "state": self.state.value,
            "category": self.category,
            "stack_frames": [frame.to_dict() for frame in self.stack_frames],
            "lock_info": self.lock_info,
            "metadata": dict(self.metadata),
            "language": self.language,
            "source_format": self.source_format,
        }


@dataclass
class ThreadDumpBundle:
    """All snapshots emitted from a single dump file."""

    snapshots: list[ThreadSnapshot]
    source_file: str
    source_format: str
    language: str
    dump_index: int = 0
    dump_label: str | None = None
    captured_at: datetime | None = None
    metadata: dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        return {
            "source_file": self.source_file,
            "source_format": self.source_format,
            "language": self.language,
            "dump_index": self.dump_index,
            "dump_label": self.dump_label,
            "captured_at": self.captured_at.isoformat() if self.captured_at else None,
            "metadata": dict(self.metadata),
            "snapshots": [snapshot.to_dict() for snapshot in self.snapshots],
        }

File: apps/engine-native/internal/models/thread_snapshot.go
// [한글] models/thread_snapshot.go — 단일 덤프 내의 스레드 1개를
// 표현하는 ThreadSnapshot 와, 한 덤프 파일 전체를 표현하는
// ThreadDumpBundle 의 정의.
//
// 데이터 흐름
//   파일 → 플러그인.Parse() → ThreadDumpBundle{Snapshots: []ThreadSnapshot}
//                          → multi/lock 분석기 입력
//
// 핵심 의도
//   • 모든 런타임의 dump 출력이 ThreadDumpBundle 한 형태로 동질화.
//   • 분석기는 더이상 "Java 인지 Go 인지" 분기하지 않음. 단지
//     ThreadState enum + StackFrame slice + LockHandle slice 만 본다.
//   • 외부에서 식별할 수 있는 메타 정보(SourceFormat, Language) 는
//     bundle 레벨로 빼두어 finding 메시지에 출처를 남길 수 있게 함.
//
// StackSignature 의 역할
//   상위 N개 프레임을 " | " 로 join 한 문자열은 멀티-덤프 분석에서
//   "같은 스레드의 같은 위치" 를 찾는 dedup 키로 사용됩니다.
//   PERSISTENT_BLOCKED_THREAD 등의 finding 이 이 시그니처로 그룹.
package models

import (
	"strings"
	"time"
)

// ThreadSnapshot is one thread captured in one dump. Mirrors
// `models/thread_snapshot.py:ThreadSnapshot`.
//
// Pointer fields (ThreadID / Category / LockInfo / Language /
// SourceFormat / LockWaiting) emit JSON `null` when unset to keep
// shape parity with Python's `Optional[...]` defaults.
type ThreadSnapshot struct {
	SnapshotID   string         `json:"snapshot_id"`
	ThreadName   string         `json:"thread_name"`
	ThreadID     *string        `json:"thread_id"`
	State        ThreadState    `json:"state"`
	Category     *string        `json:"category"`
	StackFrames  []StackFrame   `json:"stack_frames"`
	LockInfo     *string        `json:"lock_info"`
	Metadata     map[string]any `json:"metadata"`
	Language     *string        `json:"language"`
	SourceFormat *string        `json:"source_format"`
	LockHolds    []LockHandle   `json:"lock_holds"`
	LockWaiting  *LockHandle    `json:"lock_waiting"`
}

// NewThreadSnapshot builds a snapshot with non-nil slices/maps so JSON
// emits `[]` / `{}` instead of `null`. The caller still has to pass
// the required identity fields (snapshot id, name, state).
func NewThreadSnapshot(snapshotID, threadName string, state ThreadState) ThreadSnapshot {
	return ThreadSnapshot{
		SnapshotID:  snapshotID,
		ThreadName:  threadName,
		State:       state,
		StackFrames: []StackFrame{},
		Metadata:    map[string]any{},
		LockHolds:   []LockHandle{},
	}
}

// DefaultStackSignatureDepth matches Python's `depth=5` default.
const DefaultStackSignatureDepth = 5

// StackSignature returns a compact representation of the top `depth`
// frames joined with `" | "`. Matches the Python convention so legacy
// dashboards keep grouping the same threads after the multi-dump
// pipeline takes over.
//
// `depth <= 0` is treated like Python's `depth=5` default, which is
// also what `stack_signature()` uses when called without args.
//
// [한글] 알고리즘
//   1) depth <= 0 → DefaultStackSignatureDepth(=5).
//   2) 스택이 비어있으면 고정 문자열 "(no-stack)" 반환 — 통계 표에
//      "스택 없음" 도 한 그룹으로 묶이도록.
//   3) 상위 depth 개 프레임에 대해 frame.Render() 를 모아서 " | " 로
//      join. depth 가 실제 프레임 수보다 크면 자동 clamp.
//
// 시그니처가 짧은 이유
//   전체 스택을 키로 쓰면 그룹이 너무 많아져 finding 의 신호가 약해
//   집니다. 상위 5 프레임만 봐도 "같은 위치에서 멈춘 스레드" 는 거의
//   같은 그룹으로 묶이고, 그 아래 프레임 차이는 dashboard 의 detail
//   화면에서 보면 됩니다.
func (s ThreadSnapshot) StackSignature(depth int) string {
	if depth <= 0 {
		depth = DefaultStackSignatureDepth
	}
	if len(s.StackFrames) == 0 {
		return "(no-stack)"
	}
	limit := depth
	if limit > len(s.StackFrames) {
		limit = len(s.StackFrames)
	}
	rendered := make([]string, 0, limit)
	for _, frame := range s.StackFrames[:limit] {
		rendered = append(rendered, frame.Render())
	}
	return strings.Join(rendered, " | ")
}

// ThreadDumpBundle carries all snapshots emitted from a single dump
// file along with provenance (dump index, captured timestamp, source
// file, detected format). Mirrors `ThreadDumpBundle` in Python.
//
// CapturedAt is `*time.Time` so missing timestamps emit `null`.
// DumpLabel is `*string` for the same reason.
type ThreadDumpBundle struct {
	Snapshots    []ThreadSnapshot `json:"snapshots"`
	SourceFile   string           `json:"source_file"`
	SourceFormat string           `json:"source_format"`
	Language     string           `json:"language"`
	DumpIndex    int              `json:"dump_index"`
	DumpLabel    *string          `json:"dump_label"`
	CapturedAt   *time.Time       `json:"captured_at"`
	Metadata     map[string]any   `json:"metadata"`
}

// NewThreadDumpBundle constructs a bundle with non-nil snapshot/
// metadata defaults, matching Python's dataclass field factories.
func NewThreadDumpBundle(sourceFile, sourceFormat, language string) ThreadDumpBundle {
	return ThreadDumpBundle{
		Snapshots:    []ThreadSnapshot{},
		SourceFile:   sourceFile,
		SourceFormat: sourceFormat,
		Language:     language,
		Metadata:     map[string]any{},
	}
}
